Tabular data parsing improvements

william-gross opened the pull request:
This is a refactoring of TabularDataParsing as well as a new feature: a list of required columns, which if missing will generate validation errors. This prevents unhandled exceptions during line processing when users submit incomplete files for import.

william-gross committed:
ParsedLine explicit implementation

Modified ParsedLine classes to use explicit member implementation when possible. Also reformatted code.

william-gross committed:
Removed obsolete CsvLineParser constructor

There was no way for client code to use it, and its code was exactly duplicated in TextBasedTabularDataParser.buildColumnHeadersToIndexesDictionary.

william-gross committed:
Simplified TextBasedTabularDataParser.Parse

It's now parseLine and just returns field data, which reduces duplication in implementations.

william-gross committed:
Made TextBasedParsedLine.lineNumber immutable

william-gross committed:
Improved TextBasedParsedLine.fields encapsulation

william-gross committed:
Made columnHeadersToIndexes immutable

william-gross committed:
Cleaned up CsvLineParser

Removed unnecessary members and reformatted.

william-gross committed:
Cleaned up FixedWidthParser

Fixed nullability warnings and reformatted.

william-gross committed:
Cleaned up TextBasedTabularDataParser

Fixed nullability warnings and reformatted.

william-gross committed:
Moved hasHeaderRow setting into TabularDataParser

william-gross committed:
Implemented TabularDataParser required columns

william-gross committed:
Annotated StringTools.IsNullOrWhiteSpace

Fixes some nullability warnings.

william-gross committed:
Implemented required columns in ExcelParser

william-gross committed:
Backported ExcelParser improvements

Backported improvements from last commit into TextBasedTabularDataParser and TextBasedParsedLine.

william-gross committed:
Fixed tests

william-gross committed:
Formatted TabularDataParser

samrueby commented:
What happens when more columns exist in the CSV than the required-columns passed?

william-gross commented:
> What happens when more columns exist in the CSV than the required-columns passed?

Nothing. Everything works just as it did before if columns aren't included; the new feature is just an additional check before beginning line processing, that adds validation errors for missing columns.

william-gross committed:
Disambiguated ValidationError classes

william-gross committed:
Reversed param order of LineProcessingMethod

THis makes it more consistent with the validation methods of EWL web-framework controls.
